### PR TITLE
[MESH] fix #36463 : display last dataset at the end of the time extent

### LINKED ISSUE #36463: Time controller for mesh 1D usability
1. Timestep of time slider should be equal or plural of the output time step, so not smaller
2. No data is given at the end of the simulation instead of the last output timestep.

### PR DESCRIPTION
**Pull request by vcloarec**
fix #36463 (point 2)
Depending of datasets, when the temporal  slider is at the end of the time extent, no datasets was displayed, and the last datasets could never be displayed.
This PR fix this issue.

### CODE REVIEW

**Comment by vcloarec** [comment]
About how the dataset has to be selected depending on the temporal controller range, after thinking about, I am not sure the actual implementation is satisfying. 
I see two other approaches than can be better. Both use only the start time of the temporal controller range, has pros/cons, but I don't know which one is the better:
- select the last dataset with time less than or equal to the start of the range.
pros : on canvas, dataset will change when the time passed the last dataset time
cons : if time of dataset have some few miliseconds or seconds more than the start of the range, the dataset will changed only for the next time step.
- select the closest dataset from the the start of the range.
pros : this is always the closest dataset that will be rendered;
cons : on canvas, dataset change will appears between to dataset time (on the middle) and not when time fo the last dataset will be passed.

**Comment by vcloarec** [comment]
To avoid this issue and to have consistent behavior, it is better to let the user choose the matching method. So, this new approach let the user choose between the two matching methods described above.
image

File: src/app/mesh/qgsmeshlayerproperties.cpp [hunk 1]
```diff
@@ -95,6 +95,11 @@ QgsMeshLayerProperties::QgsMeshLayerProperties( QgsMapLayer *lyr, QgsMapCanvas *
   delete mOptsPage_3DView;  // removes both the "3d view" list item and its page
 #endif
 
+  mComboBoxTemporalDatasetMatchingMethod->addItem( tr( "Find closest dataset before requested time" ),
```
Comment (nyalldawson) [inline src/app/mesh/qgsmeshlayerproperties.cpp:98]: HIG convention is title case for combo box items (admittedly we aren't 100% on this, but are moving closer to it)

File: src/ui/mesh/qgsmeshlayerpropertiesbase.ui [hunk 2]
```diff
@@ -671,26 +677,30 @@ border-radius: 2px;</string>
...
-            </spacer>
+            </widget>
            </item>
-           <item row="1" column="2">
-            <widget class="QLabel" name="label_6">
+           <item row="2" column="2">
+            <widget class="QLabel" name="label_10">
              <property name="text">
-              <string>Layer Start Time</string>
+              <string>Matching dataset Method</string>
```
Comment (PeterPetrik) [inline src/ui/mesh/qgsmeshlayerpropertiesbase.ui:697]: Dataset Matching Method

File: src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h [hunk 3]
```diff
@@ -34,21 +34,37 @@ class CORE_EXPORT QgsMeshDataProviderTemporalCapabilities: public QgsDataProvide
 {
   public:
 
+    /**
+     * Method to use when requesting a tempral dataset from a time
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:38]: typo: temporal dataset

Comment (PeterPetrik) [reply src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:38]: I would clarify the docs, for example: Method for selection of mesh dataset index for time interval from temporal framework

File: src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h [hunk 4]
```diff
@@ -34,21 +34,37 @@ class CORE_EXPORT QgsMeshDataProviderTemporalCapabilities: public QgsDataProvide
...
+     **/
+    enum MatchingTemporalDatasetMethod
+    {
+      FindClosestDatasetBeforeStartRangeTime, //! Find the closest dataset which have its time before the requested start range time
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:42]: docs: Finds

File: src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h [hunk 5]
```diff
@@ -34,21 +34,37 @@ class CORE_EXPORT QgsMeshDataProviderTemporalCapabilities: public QgsDataProvide
...
+      FindClosestDatasetFromStartRangeTime //! Find the closest dataset before or after the requested start range time
```
Comment (PeterPetrik) [inline src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h:43]: Finds

**Comment by PeterPetrik** [comment]
the windows test failure seems unrelated, let me restart it

**Comment by PeterPetrik** [comment]
i have absolutely no idea how that windows failure https://cdash.orfeo-toolbox.org/testDetails.php?test=19522666&build=42496 is related to this change 
@nyalldawson do you know if this  PyQgsLocalizedDataPathRegistry (Failed)  is some random failure/broken test?

**Review by PeterPetrik — APPROVED** [review]

**Comment by nyalldawson** [comment]
It's unrelated

**Pull request merged**